### ISSUE TITLE
feat: FA → vLLM → SGLang

### ISSUE THREAD

**Issue by tuanlda78202** [tuanlda78202/gpt-oss-amd#21]
## Flowchart

```mermaid
flowchart TD
  %% ---------- Client / Ingress ----------
  subgraph Client_Ingress
    U[Client Apps]
    API[HTTP or gRPC API]
    U --> API
  end

  %% ---------- SGLang Layer ----------
  subgraph SGLang
    Q[Admission Queue]
    CB[Continuous Batching Scheduler]
    RAD[Radix Tree Cache]
    RT[Cache Aware Router]
    PLAN[Execution Planner]
    API --> Q --> CB
    CB <--> RAD
    CB --> RT
    CB -->|prefill microbatches| PLAN
    CB -->|per step decode Q_t| PLAN
    RAD -->|prefix hit: return page_ids| PLAN
  end

  %% ---------- Worker (per GPU) ----------
  subgraph Worker_GPU
    %% vLLM PagedAttention KV manager
    subgraph KV_Manager
      BT[Block Table per Sequence]
      COW[Copy on Write and Refcounts]
      PAGES[GPU KV Pages: n_kv_heads x B x head_dim]
      TIER[Optional CPU Pinned Pages]
      EVICT[Eviction Policy: LRU and refcounts]
      BT --- COW
      BT --- PAGES
      PAGES <--> TIER
      EVICT --- BT
    end

    %% FlashAttention kernels
    subgraph FlashAttention_Kernels
      PREF[Prefill Kernel FA2 or FA3]
      DEC[Decode Kernel FA Streaming Softmax]
    end

    %% Model compute
    subgraph Transformer_Compute
      EMB[Embeddings and RoPE]
      BLK[Decoder Blocks and MLPs]
      LMH[LM Head and Sampler]
    end
  end

  %% ---------- Multi-GPU ----------
  subgraph Multi_GPU
    TP[Tensor or Context Parallel Shards]
    P2P[P2P or NCCL or HIP P2P]
    TP --- P2P
  end

  %% ---------- Dataflow wiring ----------
  RT -->|route batch to best worker| Worker_GPU
  PLAN -->|build page gather lists| PREF
  PLAN --> DEC

  %% Prefill path
  CB -->|token ids| EMB
  EMB --> PREF
  PREF -->|append K and V| PAGES
  PREF -->|update tables| BT
  PREF --> BLK

  %% Decode path
  CB -->|Q_t per step| DEC
  DEC -->|gather by page_ids| PAGES
  DEC --> BLK

  %% Model head and feedback
  BLK --> LMH -->|next token or logits| CB
  LMH -->|stream tokens| API

  %% Multi-GPU hookups
  PREF --- TP
  DEC --- TP
  PAGES --- P2P

  classDef box fill:#f6f8fa,stroke:#b8c2cc,rx:6,ry:6
  class Client_Ingress,SGLang,Worker_GPU,KV_Manager,FlashAttention_Kernels,Transformer_Compute,Multi_GPU box
```

## Sequence Diagram

```mermaid
sequenceDiagram
  autonumber
  participant U as Client
  participant API as API
  participant S as SGLang Scheduler
  participant X as Radix Cache
  participant P as Planner
  participant V as vLLM Paged KV
  participant K as FlashAttention Kernels
  participant M as Model and Sampler

  U->>API: Submit prompts
  API->>S: Enqueue requests
  S->>X: Prefix lookup
  alt Cache hit
    X-->>S: page_ids alias
  else Cache miss
    S->>P: Make prefill plan
    P->>K: Prefill batch FA2 or FA3
    K->>V: Write KV pages and update block tables (copy on write)
    S->>X: Register prefix in radix tree
  end

  loop Decode steps t = 1 to T
    S->>P: Make decode plan
    P->>K: Decode with FA streaming softmax over pages
    K->>V: Gather KV via block table
    K-->>M: Attention outputs
    M-->>S: Next token logits
    S->>X: Update reuse stats
    par Optional tiering or eviction
      S->>V: Move cold pages to CPU or evict
    end
  end

  S-->>API: Stream tokens
  API-->>U: Tokens and final response
```
 
* **FlashAttention (FA)** → the **kernel** that computes attention with tiled loads + online softmax (fast + exact). ([[arXiv](https://arxiv.org/pdf/2307.08691?utm_source=chatgpt.com)][1], [[Tri Dao](https://tridao.me/blog/2024/flash3/?utm_source=chatgpt.com)][2], [[Courses CS Washington](https://courses.cs.washington.edu/courses/cse599m/23sp/notes/flashattn.pdf?utm_source=chatgpt.com)][3])
* **vLLM / PagedAttention** → the **KV-cache memory manager** (paged blocks + block table + refcounts + copy-on-write) that unlocks huge batch sizes and dynamic sequence growth. ([[vLLM Blog](https://blog.vllm.ai/2023/06/20/vllm.html?utm_source=chatgpt.com)][4], [[arXiv](https://arxiv.org/pdf/2309.06180?utm_source=chatgpt.com)][5])
* **SGLang / RadixAttention** → the **prefix-reuse layer + scheduler** (radix-tree of cached prompts/results, cache-aware routing, hierarchical KV). It prevents recompute across requests and workflows. ([[lmsys.org](https://lmsys.org/blog/2024-01-17-sglang/?utm_source=chatgpt.com)][6], [[arXiv](https://arxiv.org/html/2312.07104v1?utm_source=chatgpt.com)][7], [[llmsystem](https://llmsystem.github.io/llmsystem2025spring/assets/files/llmsys-25-sglang-72edc5043338f59db34d47e5b96ac870.pdf?utm_source=chatgpt.com)][8])

---

### 0) Mental model (layers)

```
[ Scheduler ]    ← continuous batching + cache-aware routing (SGLang)
    ↓
[ Reuse layer ]  ← Radix tree of prefixes (RadixAttention)
    ↓
[ KV manager ]   ← paged KV, block table, CoW (vLLM PagedAttention)
    ↓
[ Kernels ]      ← tiled attention (FlashAttention) over paged K/V
```

### 1) Data structures

* **Pages**: fixed-size KV blocks of `B` tokens, layout `[n_kv_heads, B, head_dim]` contiguous in `head_dim`.
* **Block table (per sequence)**: `[(page_id, fill_count), ...]`. Implement **refcount + copy-on-write** so diverging sequences can share pages until a write. ([[vLLM Blog](https://blog.vllm.ai/2023/06/20/vllm.html?utm_source=chatgpt.com)][4])
* **Radix tree**: nodes hold a *vector of page\_ids* for the prefix. Leaves point to current sequences. LRU metadata supports eviction; keep a hit counter for scheduling. ([[lmsys.org](https://lmsys.org/blog/2024-01-17-sglang/?utm_source=chatgpt.com)][6], [[arXiv](https://arxiv.org/html/2312.07104v1?utm_source=chatgpt.com)][7])

### 2) Prefill (bulk token ingest)

* Run an **FA2 prefill** kernel on a microbatch (contiguous matrices → maximum arithmetic intensity).
* **Append** K/V of each request into paged cache; when multiple requests share a prefix, **alias** the same pages via the block table (refcount++). This is vLLM’s CoW in practice. ([[arXiv](https://arxiv.org/pdf/2309.06180?utm_source=chatgpt.com)][5])
* Insert the page list into the **radix tree** so future requests can reuse that KV instantly. ([[lmsys.org](https://lmsys.org/blog/2024-01-17-sglang/?utm_source=chatgpt.com)][6])

### 3) Decode (one step per request)

* For each query head, perform **FlashAttention’s streaming softmax across pages**:

  1. Follow the sequence’s block table; for each page, load a `K/V` tile into LDS (shared).
  2. Compute scores `Q·Kᵀ`, update `(m, l, o)` using FA’s running-max trick, then move to the next page.
  3. Final `o / l` is the head output.
     Only change vs classic FA: your **tile loader is “paged”** (indirection via block table). ([[arXiv](https://arxiv.org/pdf/2307.08691?utm_source=chatgpt.com)][1])

### 4) Scheduling (keep GPUs full)

* Implement **continuous batching**: admit/retire requests every step; group by similar “remaining length”. This keeps utilization high under skewed lengths. ([[Anyscale](https://www.anyscale.com/blog/continuous-batching-llm-inference?utm_source=chatgpt.com)][9])
* **Cache-aware routing** (SGLang): if a new request hits a long prefix in the radix tree, collocate it on the worker that holds those pages to maximize reuse and avoid cross-device copies. ([[llmsystem](https://llmsystem.github.io/llmsystem2025spring/assets/files/llmsys-25-sglang-72edc5043338f59db34d47e5b96ac870.pdf?utm_source=chatgpt.com)][8])

### 5) Eviction & memory tiers

* **PagedAttention eviction**: drop least-recent pages by refcount/LRU.
* **Radix tree eviction**: evict cold subtrees; detach their page lists (decrement refcounts).
* If needed, add **hierarchical KV** (GPU→CPU) for very long contexts; bring hot pages back on demand. (SGLang design mentions hierarchical KV.) ([[llmsystem](https://llmsystem.github.io/llmsystem2025spring/assets/files/llmsys-25-sglang-72edc5043338f59db34d47e5b96ac870.pdf?utm_source=chatgpt.com)][8])

---

## AMD MI250X (ROCm/HIP) notes

* FA2 backends exist on ROCm (CK or Triton). If you’re rolling your own, match the FA2 tiling pattern and wavefront-friendly reductions (64-lane). Keep LDS usage under \~64 KB/CU per tile. ([[rocm.docs.amd.com](https://rocm.docs.amd.com/en/latest/how-to/rocm-for-ai/inference-optimization/model-acceleration-libraries.html?utm_source=chatgpt.com)][10])
* If you later test against vLLM/FA on ROCm, note: MI210/MI250/MI300 use CK; Triton support varies by arch; NAVI GPUs differ. ([[GitHub](https://github.com/vllm-project/vllm/issues/4514?utm_source=chatgpt.com)][11])

---

## Pitfalls & tips

* **Page size (`B`)**: too small → poor bandwidth; too large → waste on tails and worse fragmentation. Start with 64–128 tokens/page; profile. (vLLM discusses block granularity impacts.) ([[vLLM Blog](https://blog.vllm.ai/2023/06/20/vllm.html?utm_source=chatgpt.com)][4])
* **Copy-on-write correctness**: when you append to a shared tail, allocate a fresh page, copy only the written slice, and flip the block table entry for *that* sequence. ([[vLLM Blog](https://blog.vllm.ai/2023/06/20/vllm.html?utm_source=chatgpt.com)][4])
* **Chunked prefill**: If inputs are huge, chunk prefill to overlap compute/append; keep FA tiles large and writes sequential per page. (Common in FlashInfer writeups.) ([[FlashInfer](https://flashinfer.ai/2024/02/02/introduce-flashinfer.html?utm_source=chatgpt.com)][12])
* **Plan→Run separation**: Precompute gather lists of `(page_id, offsets)` for each sequence (“plan”), then launch FA tiles accordingly (“run”) to reduce divergence. FlashInfer uses this idea for load-balancing. ([[GitHub](https://github.com/flashinfer-ai/flashinfer?utm_source=chatgpt.com)][13])

---

### TL;DR implementation recipe

1. **FA2 kernels** for prefill & decode. ([[arXiv](https://arxiv.org/pdf/2307.08691?utm_source=chatgpt.com)][1])
2. **Paged KV + block table + CoW** (vLLM). ([[arXiv](https://arxiv.org/pdf/2309.06180?utm_source=chatgpt.com)][5])
3. **Radix tree** to persist/reuse KV across requests (SGLang). ([[lmsys.org](https://lmsys.org/blog/2024-01-17-sglang/?utm_source=chatgpt.com)][6])
4. **Continuous batching + cache-aware routing** to keep SMs/CUs busy. ([[Anyscale](https://www.anyscale.com/blog/continuous-batching-llm-inference?utm_source=chatgpt.com)][9], [[llmsystem](https://llmsystem.github.io/llmsystem2025spring/assets/files/llmsys-25-sglang-72edc5043338f59db34d47e5b96ac870.pdf?utm_source=chatgpt.com)][8])

[1]: https://arxiv.org/pdf/2307.08691?utm_source=chatgpt.com "FlashAttention-2: Faster Attention with Better Parallelism ..."
[2]: https://tridao.me/blog/2024/flash3/?utm_source=chatgpt.com "FlashAttention-3: Fast and Accurate Attention with ..."
[3]: https://courses.cs.washington.edu/courses/cse599m/23sp/notes/flashattn.pdf?utm_source=chatgpt.com "From Online Softmax to FlashAttention"
[4]: https://blog.vllm.ai/2023/06/20/vllm.html?utm_source=chatgpt.com "vLLM: Easy, Fast, and Cheap LLM Serving with PagedAttention"
[5]: https://arxiv.org/pdf/2309.06180?utm_source=chatgpt.com "Efficient Memory Management for Large Language Model ..."
[6]: https://lmsys.org/blog/2024-01-17-sglang/?utm_source=chatgpt.com "Fast and Expressive LLM Inference with RadixAttention ..."
[7]: https://arxiv.org/html/2312.07104v1?utm_source=chatgpt.com "Efficiently Programming Large Language Models using ..."
[8]: https://llmsystem.github.io/llmsystem2025spring/assets/files/llmsys-25-sglang-72edc5043338f59db34d47e5b96ac870.pdf?utm_source=chatgpt.com "Efficient LLM Inference with SGLang"
[9]: https://www.anyscale.com/blog/continuous-batching-llm-inference?utm_source=chatgpt.com "Achieve 23x LLM Inference Throughput & Reduce p50 ..."
[10]: https://rocm.docs.amd.com/en/latest/how-to/rocm-for-ai/inference-optimization/model-acceleration-libraries.html?utm_source=chatgpt.com "Model acceleration libraries - ROCm Documentation - AMD"
[11]: https://github.com/vllm-project/vllm/issues/4514?utm_source=chatgpt.com "[Bug]: For RDNA3 (navi31; gfx1100) ..."
[12]: https://flashinfer.ai/2024/02/02/introduce-flashinfer.html?utm_source=chatgpt.com "Accelerating Self-Attentions for LLM Serving with FlashInfer"
[13]: https://github.com/flashinfer-ai/flashinfer?utm_source=chatgpt.com "FlashInfer: Kernel Library for LLM Serving"
[14]: https://fergus.hashnode.dev/attention-kernels-for-llm-inference?utm_source=chatgpt.com "Attention kernels for LLM inference - fergus"
